tokio-util/src/sync/cancellation_token/implementation.rs

```diff
@@ -0,0 +1,341 @@
+        locked_node.num_handles
+    };
+
+    if num_handles == 0 {
+        with_locked_node_and_parent(node, |mut node, parent| {
+            // Remove the node from the tree
+            match parent {
+                Some(mut parent) => {
+                    // As we want to remove ourselves from the tree,
+                    // we have to move the children to the parent, so that
+                    // they still receive the cancellation event without us.
+                    // Moving them does not violate invariant #1.
+                    move_children_to_parent(&mut node, &mut parent);
+
+                    // Remove the node from the parent
+                    remove_child(&mut parent, node);
+                }
+                None => {
+                    // Due to invariant #1, we can assume that our
+                    // children can no longer be cancelled through us.
+                    // (as we now have neither a parent nor handles)
+                    // Therefore we can disconnect them.
+                    disconnect_children(&mut node);
+                }
+            }
+        });
+    }
+}
+
+/// Cancels a node and its children.
+///
+/// Then, dissolves the entire subtree for easier disposal.
+///
+/// Invariant #1:
+/// It's important that during the entire process of cancellation, there is
+/// never a gap in lock propagation. We need to lock both the parent and the child for the
+/// entire duration of cancellation and disconnection. Those cannot happen separately,
+/// otherwise other calls like [decrease_handle_refcount] could modify the tree structure
+/// while we recurse over it.
+///
+/// We basically have to create a "lock wave" from the parent to children, that nothing else
+/// can bypass. Otherwise, other actions could add/remove children that our cancellation would miss.
+///
+/// Example: Holding a reference of a child does not influence its handle reference count, so
+/// it will happily drop its grandchildren in [decrease_handle_refcount] before we got the chance
+/// to propagate the cancellation to them, if we violate invariant #1.
+/// Unless we keep the child locked during the entire time, from the moment on when we
+/// disconnected it from its parent.
+pub(crate) fn cancel(node: &Arc<TreeNode>) {
+    let waker = &node.waker;
+
+    with_locked_node_and_parent(node, |mut node, parent| {
+        // Cancel the subtree
+        node.is_cancelled = true;
+        waker.notify_waiters();
+
+        // One by one, adopt grandchildren and then cancel and detach the child
+        while let Some(child) = node.children.pop() {
+            // `node` is already locked, so we can lock the child
+            let mut locked_child = child.inner.lock().unwrap();
+
+            // Detach the child from node
+            // No need to modify node.children, as the child already got removed with `.pop`
+            locked_child.parent = None;
+            locked_child.parent_idx = 0;
```

Comment:
If the node is already cancelled, you can just `continue` after these two statements.